Source/WebCore/platform/graphics/texmap/TextureMapperLayer3DRenderingContext.cpp

```diff
@@ -29,118 +29,277 @@
+        return aMin < bMin;
     });
+
+    SweepAndPrunePairs potentialIntersectionPairs;
+    for (size_t i = 0; i < indices.size(); ++i) {
+        for (size_t j = i + 1; j < indices.size(); ++j) {
+            auto firstIndex = indices[i];
+            auto secondIndex = indices[j];
+
+            auto [iMin, iMax] = layerBounds(layers[firstIndex], axis);
+            auto [jMin, jMax] = layerBounds(layers[secondIndex], axis);
+            if (jMin >= iMax)
+                break; // No overlap possible
+
+            // Ensure canonical order (smaller index first)
+            if (firstIndex > secondIndex)
+                std::swap(firstIndex, secondIndex);
+
+            potentialIntersectionPairs.insert({ firstIndex, secondIndex });
+        }
+    }
+    return potentialIntersectionPairs;
+}
+
+FloatQuad TextureMapperLayer3DRenderingContext::projectLayerToZYPlane(const Layer& layer) const
+{
+    auto p1 = layer.geometry.vertexAt(0);
+    auto p2 = layer.geometry.vertexAt(1);
+    auto p3 = layer.geometry.vertexAt(2);
+    auto p4 = layer.geometry.vertexAt(3);
+    return FloatQuad(
+        { p1.z(), p1.y() },
+        { p2.z(), p2.y() },
+        { p3.z(), p3.y() },
+        { p4.z(), p4.y() });
+}
+
+// Given two points defining an edge, returns the perpendicular axis (normalized)
+FloatPoint TextureMapperLayer3DRenderingContext::edgeNormal(const FloatPoint& p1, const FloatPoint& p2) const
+{
+    auto edge = p2 - p1;
+
+    // A perpendicular to (x,y) is (y,-x)
+    FloatPoint normal = { edge.height(), -edge.width() };
+    normal.normalize();
+
+    return normal;
+}
+
+// Project a convex polygon onto an axis and return min/max scalar values
+std::pair<float, float> TextureMapperLayer3DRenderingContext::projectQuadOnAxis(const FloatQuad& quad, const FloatPoint& axis) const
+{
+    float p1 = quad.p1().dot(axis);
+    float p2 = quad.p2().dot(axis);
+    float p3 = quad.p3().dot(axis);
+    float p4 = quad.p4().dot(axis);
+
+    float min = min4(p1, p2, p3, p4);
+    float max = max4(p1, p2, p3, p4);
+
+    return { min, max };
 }
 
-// Build BSP tree for rendering polygons with painter's algorithm.
+// Intersection check using Separating Axis Theorem
+// For more information:
+// https://en.wikipedia.org/wiki/Hyperplane_separation_theorem
+bool TextureMapperLayer3DRenderingContext::quadsIntersect(const FloatQuad& quadA, const FloatQuad& quadB) const
+{
+    FloatPoint axes[8];
+
+    // QuadA edges: (1->2), (2->3), (3->4), (4->0)
+    axes[0] = edgeNormal(quadA.p1(), quadA.p2());
+    axes[1] = edgeNormal(quadA.p2(), quadA.p3());
+    axes[2] = edgeNormal(quadA.p3(), quadA.p4());
+    axes[3] = edgeNormal(quadA.p4(), quadA.p1());
+
+    // QuadB edges: (1->2), (2->3), (3->4), (4->0)
+    axes[4] = edgeNormal(quadB.p1(), quadB.p2());
+    axes[5] = edgeNormal(quadB.p2(), quadB.p3());
+    axes[6] = edgeNormal(quadB.p3(), quadB.p4());
+    axes[7] = edgeNormal(quadB.p4(), quadB.p1());
+
+    for (int i = 0; i < 8; ++i) {
```

Comment:
How about `for (auto& axis : axes) {` or `std::all_of`?

Reply:
Fixed